Framework/API/inc/MantidAPI/Algorithm.h

```diff
@@ -75,16 +78,16 @@ Gaudi user guide).
 http://proj-gaudi.web.cern.ch/proj-gaudi/)
 @date 12/09/2007
 */
-class MANTID_API_DLL Algorithm : public IAlgorithm, public Kernel::PropertyManagerOwner {
```

Comment:
I like the fact we've dropped the diamond inheritance, I'm genuinely surprised someone never ran into relying on the override in the weaker class.

But why do we now return `Algorithm` instead of `IAlgorithm` for the getters now? Since were still conforming to the base virtual class and this removes some flexibility when it comes to mocking. 
Admittedly we don't have many (if any) tests that mock out an algorithm. But this also removes the flexibility to do so too hence my confusion

Reply:
> I like the fact we've dropped the diamond inheritance, I'm genuinely surprised someone never ran into relying on the override in the weaker class.
> 
> But why do we now return `Algorithm` instead of `IAlgorithm` for the getters now? Since were still conforming to the base virtual class and this removes some flexibility when it comes to mocking.
> Admittedly we don't have many (if any) tests that mock out an algorithm. But this also removes the flexibility to do so too hence my confusion

I think in an earlier version of this change I had done a more significant change to the class hierarchy where IAlgorithm was no longer inhering from IPropertyManager. Martyn had suggested that having one interface inheriting from another might not be ideal. However I backtracked from that because the size of the change in terms of number of files impacted was huge - I had to replace IAlgorithm with Algorithm in loads of places. Perhaps I missed one or two though so I think I can revert the changes to the methods fromString, fromJson, fromHistory +  the changes on the AlgorithmNotification. The fromXXX are all static so not sure how easy it would be to mock them but since I think these changes weren't meant to be included I'll revert them anyway.